src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs

```diff
@@ -327,10 +327,19 @@ internal static void ApplyParameterInfo(this JsonNode schema, ApiParameterDescri
     /// <param name="createSchemaReferenceId">A delegate that generates the reference ID to create for a type.</param>
     internal static void ApplyPolymorphismOptions(this JsonNode schema, JsonSchemaExporterContext context, Func<JsonTypeInfo, string?> createSchemaReferenceId)
     {
-        // The `context.Path.Length == 0` check is used to ensure that we only apply the polymorphism options
+        // The `context.BaseTypeInfo == null` check is used to ensure that we only apply the polymorphism options
         // to the top-level schema and not to any nested schemas that are generated.
-        if (context.TypeInfo.PolymorphismOptions is { } polymorphismOptions && context.Path.Length == 0)
+        if (context.TypeInfo.PolymorphismOptions is { } polymorphismOptions && context.BaseTypeInfo == null)
         {
+            // System.Text.Json supports serializing to a non-abstract base class if no discriminator is provided.
+            // OpenAPI requires that all polymorphic sub-schemas have an associated discriminator. To maintain accuracy,
+            // with the actual implementation and comply with the limitations of the OpenAPI schema we do not emit
+            // a `discriminator` property for types that are represented this way.
+            if (!context.TypeInfo.Type.IsAbstract &&
+                !polymorphismOptions.DerivedTypes.Any(type => type.DerivedType == context.TypeInfo.Type))
```

Comment:
Isn't the interesting case when the non-abstract class is the base type?  I might be mixed up.  Why are we checking whether it's the derived type?

Reply:
`PolymorphismOptions.DerivedTypes` is populated from the `JsonDerivedTypeAttribute` annotations on a type, which provide a way for defining what discriminator is assocaited with a type.  We care when there is no derived type assocated with the base type because then there is no discriminator defined.

Essentially, we're checking for the difference between:

```
// Scenario A:
[JsonDerivedType(typeof(Cat), typeDiscriminator: "cat")]
[JsonDerivedType(typeof(Dog), typeDiscriminator: "dog")]
[JsonDerivedType(typeof(Pet), typeDiscriminator: "pet")]
internal class Pet { } 
```

and

```
// Scenario B:
[JsonDerivedType(typeof(Cat), typeDiscriminator: "cat")]
[JsonDerivedType(typeof(Dog), typeDiscriminator: "dog")]
internal class Pet { } 
```

In Scenario A, there's a discriminator associated with the `Pet` type so we can use it in compliance with OpenAPI's expectations. In Scenario B, there's no discriminator assocaited with `Pet` type but STJ will still  deserialize to it if no `$type` property is provided in a JSON object.

Reply:
"If the base type doesn't declare itself as its own derived type, then it can't have specified a discriminator, which OpenAPI requires."?

Reply:
Updated.